### PR TITLE
Better error message when a bad method type is provided

### PR DESCRIPTION
**Pull request by mscheltienne**
Better error message raised when someone accidentally provided the `method` argument as a `str`, e.g.

```
from mne.datasets import sample
from mne.io import read_raw_fif


directory = sample.data_path() / "MEG" / "sample"
raw = read_raw_fif(directory / "sample_audvis_raw.fif", preload=False)
raw.pick("eeg", exclude=[]).load_data()
raw.interpolate_bads(method="spline")
```

On main:

```
ValueError: Invalid value for the 'method['meg']' parameter. Allowed values are 'MNE' and 'nan', but got 'spline' instead.
```

On this PR:

```
TypeError: method must be an instance of dict or None, got <class 'str'> instead.
```

### CODE REVIEW

**Comment by larsoner** [comment]
`str` should be allowed when there is only one channel type. So what you wrote should have worked. The internal logic must be broken because internally it should have checked for one channel type, found eeg, and upconverted to `dict(eeg='spline')`

**Comment by mscheltienne** [comment]
Yes, that would be nicer.. but then yes the logic must be changed to check for single-channel type objects. Will do later.

**Comment by larsoner** [comment]
It doesn't even have to be single channel type thinking about it more. method='mne' or 'nan' should work even if meg and eeg are present for example. So really we should just upconvert to dict only for data type keys that are present in the data. In this case that would just be eeg. And for M/EEG if you pass 'spline' you'd get error you saw, which would be correct (enough). The bug here I think is that we probably create a dict with too many keys, I.e., add the meg key even though no meg channels are present anymore

**Comment by mscheltienne** [comment]
@larsoner Ready for review/merge.

File: mne/channels/channels.py [hunk 1]
```diff
@@ -828,24 +829,25 @@ def interpolate_bads(
...
-            All channel types support "nan".
-            The key ``"eeg"`` has two additional options:
 
-            - ``"spline"`` (default)
-                Use spherical spline interpolation.
-            - ``"MNE"``
-                Use minimum-norm projection to a sphere and back.
-                This is the method used for MEG channels.
+            - ``"meg"`` channels support ``"MNE"`` (default) and ``"nan"``
+            - ``"eeg"`` channels support ``"spline" (default), ``"MNE"`` and ``"nan"``
```
Comment (mscheltienne) [inline mne/channels/channels.py:836]:
```suggestion
            - ``"eeg"`` channels support ``"spline"`` (default), ``"MNE"`` and ``"nan"``
```

**Review by larsoner — COMMENTED** [review]
I think all that is missing is a tiny test case equivalent of the minimal example at the top

File: mne/channels/channels.py [hunk 2]
```diff
@@ -828,24 +829,25 @@ def interpolate_bads(
...
 
-            The default value for ``"meg"`` is ``"MNE"``, and the default value
-            for ``"fnirs"`` is ``"nearest"``.
-
-            The default (None) is thus an alias for::
+            None is an alias for::
 
                 method=dict(meg="MNE", eeg="spline", fnirs="nearest")
 
+            If a :class:`str` is provided, the method will be applied to all channels
```
Comment (drammock) [inline mne/channels/channels.py:843]:
```suggestion
            If a :class:`str` is provided, the method will be applied to all channel
```

File: mne/channels/channels.py [hunk 3]
```diff
@@ -872,49 +872,53 @@ def interpolate_bads(
...
         _check_preload(self, "interpolation")
+        _validate_type(method, (dict, str, None), "method")
         method = _handle_default("interpolation_method", method)
+        ch_types = self.get_channel_types(unique=True)
+        # figure out if we have "mag" for "meg", "hbo" for "fnirs", ... to filter the
+        # "method" dictionary and keep only keys that correspond to existing channels.
+        for ch_type in ("meg", "fnirs"):
+            for sub_ch_type in _second_rules[ch_type][1].values():
+                if sub_ch_type in ch_types:
+                    del ch_types[ch_types.index(sub_ch_type)]
```
Comment (drammock) [inline mne/channels/channels.py:883]:
```suggestion
                    ch_types.remove(sub_ch_type)
```
equivalent I think?

File: mne/channels/channels.py [hunk 4]
```diff
@@ -872,49 +872,53 @@ def interpolate_bads(
...
+                    if ch_type not in ch_types:
+                        ch_types.append(ch_type)
+        keys2delete = set(method) - set(ch_types)
+        for key in keys2delete:
+            del method[key]
+        logger.info("Setting channel interpolation method to %s.", method)
```
Comment (drammock) [inline mne/channels/channels.py:889]: seems odd to log this *before* checking valids

File: mne/channels/channels.py [hunk 5]
```diff
@@ -872,49 +872,53 @@ def interpolate_bads(
...
+            _check_option(f"method['{key}']", method[key], valids[key])
+        idx = _picks_to_idx(self.info, list(method), exclude=(), allow_empty=True)
+        if idx.size == 0 or len(pick_info(self.info, idx)["bads"]) == 0:
             warn("No bad channels to interpolate. Doing nothing...")
             return self
-        logger.info("Interpolating bad channels")
+        logger.info("Interpolating bad channels.")
         origin = _check_origin(origin, self.info)
-        if method["eeg"] == "spline":
+        if "eeg" in method and method["eeg"] == "spline":
```
Comment (drammock) [inline mne/channels/channels.py:904]:
```suggestion
        if method.get("eeg", "") == "spline":
```

**Review by drammock — APPROVED** [review]
Thanks for the improvement @mscheltienne! just a few small nitpicks.

**Pull request merged**